### ISSUE TITLE
File uploading being cancelled on some files

### ISSUE THREAD

**Issue by NRauh** [balderdashy/skipper#10]
I'm trying to use Skipper to save uploaded images, but it's only working on some files.
Some files finish, and successfully write to the disk. Some files finish uploading, but the file that's written is empty, and some files just hang on uploading until it times out.
When it cancels or hangs, it's on the same file, so it isn't randomly not working.

I found two working and non-working files to log the callback, and this is what it said
(these weren't uploaded all at once, and the working-not-working pattern isn't a coincidence)

```
[ { size: 2005363,
    type: 'image/jpeg',
    filename: 'Success-1.jpg',
    status: 'bufferingOrWriting',
    field: 'avatarUpload',
    extra: undefined } ]
[ { size: 251156,
    type: 'image/jpeg',
    filename: 'Cancelled-1.jpg',
    status: 'cancelled',
    field: 'avatarUpload',
    extra: undefined } ]
[ { size: 2337904,
    type: 'image/jpeg',
    filename: 'Success-2.jpg',
    status: 'bufferingOrWriting',
    field: 'avatarUpload',
    extra: undefined } ]
[ { size: 329172,
    type: 'image/jpeg',
    filename: 'Cancelled-2.jpg',
    status: 'cancelled',
    field: 'avatarUpload',
    extra: undefined } ]
```

It's also probably worth saying that if I don't comment out <code>stream.pause()</code> on line 33 of <code>node_modules/skipper/node_modules/connect/node_modules/raw-data/index.js</code> that I get this error.
```
Error: Cannot switch to old mode now.
    at emitDataEvents (_stream_readable.js:730:11)
    at IncomingMessage.Readable.pause (_stream_readable.js:721:3)
    at module.exports (/Users/Nate/Desktop/Projects/Active/project/node_modules/skipper/node_modules/connect/node_modules/raw-body/index.js:33:14)
    at json (/Users/Nate/Desktop/Projects/Active/project/node_modules/skipper/node_modules/connect/lib/middleware/json.js:51:5)
    at _parseHTTPBody (/Users/Nate/Desktop/Projects/Active/project/node_modules/skipper/index.js:71:5)
    at Object.retryBodyParser [as handle] (/Users/Nate/Desktop/Projects/Active/project/node_modules/sails/lib/express/bodyParserRetry.js:36:36)
    at next (/Users/Nate/Desktop/Projects/Active/project/node_modules/sails/node_modules/express/node_modules/connect/lib/proto.js:190:15)
    at next (/Users/Nate/Desktop/Projects/Active/project/node_modules/sails/node_modules/express/node_modules/connect/lib/proto.js:192:9)
    at Parser.next (/Users/Nate/Desktop/Projects/project/node_modules/skipper/index.js:97:20)
    at passControlToApp (/Users/Nate/Desktop/Projects/Active/project/node_modules/skipper/lib/Parser/parse.js:172:11)
    at null._onTimeout (/Users/Nate/Desktop/Projects/Active/project/node_modules/skipper/node_modules/async/lib/async.js:668:17)
    at Timer.listOnTimeout [as ontimeout] (timers.js:110:15)
```
This only happens on the files that would otherwise be successful. If I upload one that'll be blank, then it works, but still writes an empty file.

I'm using Skipper 0.1.6 with Sails 0.9.8.
But I've tried with Sails 0.9.16, with the same results

**Comment by mikermcneil** [comment]
@ArVan would you mind taking 0.5.5 for a spin?  In your sample app, you can do:

```
DEBUG=skipper,skipper:* sails lift
```

to get Skipper to spit out verbose logs.

Thanks a lot, and I appreciate your patience.  We made a lot of progress over the last few days.

**Comment by jessearmand** [comment]
I have a log from skipper 0.5.5, the issue maybe related to this, but it's more related to #49 

    Sun, 08 Mar 2015 14:06:44 GMT skipper Set up "maxTimeToWaitForFirstFile" timer for 10000ms
    Acquiring new Upstream for field `avatar`
    Sun, 08 Mar 2015 14:06:44 GMT skipper passed control to app because first file was received
    Sun, 08 Mar 2015 14:06:44 GMT skipper waiting for any text params
    Upstream: Pumping incoming file through field `avatar`
    Sun, 08 Mar 2015 14:06:44 GMT skipper Something is trying to read from Upstream `avatar`...
    Form: emitted `close`
    Sun, 08 Mar 2015 14:06:44 GMT skipper multiparty form closed.
    Sun, 08 Mar 2015 14:06:44 GMT skipper Upstream: No more files will be sent through field `avatar`- clearing timeouts...
    Sun, 08 Mar 2015 14:06:44 GMT skipper Passing control to app...
    Sun, 08 Mar 2015 14:06:44 GMT skipper .upload() called on upstream
    RenamerPump:
    • dirname => undefined
    • field => avatar
    • fd => 06a8fb6c-3e01-44b5-b641-381c288c9aa7.png
    A receiver is finished writing files from Upstream `avatar`.
    (this doesn't necessarily mean any files were actually written...)
    { fd: '06a8fb6c-3e01-44b5-b641-381c288c9aa7.png',
      size: 90487,
      type: 'image/png',
      filename: 'logo.png',
      status: 'bufferingOrWriting',
      field: 'avatar',
      extra:
       { Location: 'https://mybucket.s3.amazonaws.com/06a8fb6c-3e01-44b5-b641-381c288c9aa7.png',
         Bucket: 'mybucket',
         Key: '06a8fb6c-3e01-44b5-b641-381c288c9aa7.png',
         ETag: '"7255fcb1f1a2a609c77891156134c442-1"',
         size: 90487 } }
    Fatal incoming error in Upstream `avatar` ::   (source or user may have cancelled the request)
    Error: Request aborted
    -----------
    undefined
    'Request aborted' - 'Error'
    Sun, 08 Mar 2015 14:06:48 GMT skipper multiparty form emitted error: [Error: Request aborted]
    -----------
    Piping the not-yet-written bytes from incoming file `logo.png` to the memory hole..
    A receiver handling Upstream `avatar` encountered a write error : [Error: Request aborted]
    Fatal incoming error in Upstream `avatar` ::   (source or user may have cancelled the request)
    Error: Request aborted
    -----------
    undefined
    'Request aborted' - 'Error'
    -----------
    Sun, 08 Mar 2015 14:06:48 GMT skipper Connected upstream: Emitting read error on incoming file `logo.png` :: [Error: Request aborted]
    Sun, 08 Mar 2015 14:06:48 GMT skipper Connected upstream: Emitting read error on incoming file `logo.png` :: [Error: Request aborted]

**Comment by Alaneor** [comment]
I have a similar issue where only the first file is successfully stored on disk and the second file is written as empty.

I attempt to upload the files with `curl`:

`curl -i -F file=@./tmpfile.txt -F file=@./key.pem localhost:3300/up`

And here is skipper's detailed logs (seems like all is well):

```
  skipper Set up "maxTimeToWaitForFirstFile" timer for 10000ms +2m
Acquiring new Upstream for field `file`
  skipper passed control to app because first file was received +1ms
  skipper waiting for any text params +0ms
Upstream: Pumping incoming file through field `file`
  skipper Something is trying to read from Upstream `file`... +0ms
  skipper Passing control to app... +6ms
  skipper .upload() called on upstream +2ms
RenamerPump:
• dirname => undefined
• field => file
• fd => /path/to/storage/f31d7f2e-7350-4ce5-90c0-d278f1f38545.txt
RenamerPump:
• dirname => undefined
• field => file
• fd => /path/to/storage/300a155c-4cb7-47ca-abf3-a3e5c8a22d70.pem
  skipper Something is trying to read from Upstream `file`... +30ms
Upstream: Pumping incoming file through field `file`
Form: emitted `close`
  skipper multiparty form closed. +2ms
  skipper Upstream: No more files will be sent through field `file`- clearing timeouts... +0ms
A receiver is finished writing files from Upstream `file`.
(this doesn't necessarily mean any files were actually written...)
```

I **do**, however, pipe the incoming streams into a second writable stream. My use case is that I need to calculate SHA-1 hashes of all the files uploaded. I do it this way:

```js
var interceptor = new stream.Writable({ objectMode: true })
    , hashers = []

// We want to calculate shasum of every file that is uploaded - To do so, we use this
// writable stream to listen for incoming file uploads
interceptor._write = function newFile (filestream, encoding, next) {
  var hasher = crypto.createHash('sha1')
  // Send the file data through the hasher but make sure the hasher is not flushed on end -
  // that would cause .digest() to be called and the shasum would become inaccessible
  filestream.pipe(hasher, { end: false })
  hashers.push(hasher) // For later retrieval

  return next()
}
// file is the object returned by req.file('fileId')
file.pipe(interceptor)

// Perform the upload
file.upload(/* ... */)
```

You can see from the example that everytime there is a new file to be uploaded, the `interceptor` object's `_write()` method will be triggered and, from there, the incoming file stream is piped into the `crypto.createHash()` stream.

I saw someone mentioning that piping the stream to something else might break the upload, but I did not fully comprehend the meaning of the words, so if there are known solutions to this, I would appreciate any pointers.

> PS. removing the piping logic resolves my problem with some files being empty.

**Comment by sailsbot** [comment]
Thanks for posting, @NRauh.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
